mm2src/mm2_main/src/lp_wallet/mnemonics_storage.rs

```diff
@@ -0,0 +1,78 @@
+use crypto::EncryptedData;
+use mm2_core::mm_ctx::MmArc;
+use mm2_err_handle::prelude::*;
+use mm2_io::fs::ensure_file_is_writable;
+
+type WalletsStorageResult<T> = Result<T, MmError<WalletsStorageError>>;
+
+#[derive(Debug, Deserialize, Display, Serialize)]
+pub enum WalletsStorageError {
+    #[display(fmt = "{} db file is not writable", path)]
+    DbFileIsNotWritable { path: String },
+    #[display(fmt = "Error writing to file: {}", _0)]
+    FsWriteError(String),
+    #[display(fmt = "Error reading from file: {}", _0)]
+    FsReadError(String),
+    #[display(fmt = "Internal error: {}", _0)]
+    Internal(String),
+}
+
+/// Saves the passphrase to a file associated with the given wallet name.
+///
+/// # Arguments
+///
+/// * `wallet_name` - The name of the wallet.
+/// * `passphrase` - The passphrase to save.
+///
+/// # Returns
+/// Result indicating success or an error.
+pub(super) async fn save_encrypted_passphrase(
+    ctx: &MmArc,
+    wallet_name: &str,
+    encrypted_passphrase_data: &EncryptedData,
+) -> WalletsStorageResult<()> {
+    let wallet_path = ctx.wallet_file_path(wallet_name);
+    ensure_file_is_writable(&wallet_path).map_to_mm(|_| WalletsStorageError::DbFileIsNotWritable {
+        path: wallet_path.display().to_string(),
+    })?;
+    mm2_io::fs::write_json(encrypted_passphrase_data, &wallet_path, true)
+        .await
+        .mm_err(|e| WalletsStorageError::FsWriteError(e.to_string()))
+}
+
+/// Reads the encrypted passphrase data from the file associated with the given wallet name.
+///
+/// This function is responsible for retrieving the encrypted passphrase data from a file.
+/// The data is expected to be in the format of `EncryptedData`, which includes
+/// all necessary components for decryption, such as the encryption algorithm, key derivation
+/// details, salts, IV, ciphertext, and HMAC tag.
```

Comment:
I opted for file-based storage for encrypted passphrase/mnemonic data to allow portability, this also aligns with the industry standards afaik (e.g. wallet.dat file for bitcoin core and for komodo daemon), please also check the usecase here https://github.com/KomodoPlatform/komodo-defi-framework/pull/2014#discussion_r1410396767 where the user can upload a seed file.
For encrypted wallet data using the seed (e.g. swap files), how to save the data is not implemented yet. I will probably add methods for both file storage and database. Will add this comment to the checklist here https://github.com/KomodoPlatform/komodo-defi-framework/pull/2014#issue-2010065894 to remember.

Reply:
Storing them in the database wouldn't change the behaviour for users I believe. You just change the way you keep it's data for mm2 (e.g., you can still upload seed files). Is there any other reason for keeping them as a file on native (non-WASM) mode ? If not, I would keep this behaviour simple and same on all supported mm2 platforms and keep the mm2 internal data in a central storage (in this case mm2 db).

Reply:
I think storing encrypted passphrase/mnemonic data in a separate file instead of the main mm2 db makes backups easier for end users. They can just backup a small(KB sized) encrypted file instead of a possibly GB sized mm2 db.

Also, it will be easier to document recovery of passphrase from backup (independent of using mm2) if it was a separate file instead of a part of a db

Reply:
> If not, I would keep this behaviour simple and same on all supported mm2 platforms and keep the mm2 internal data in a central storage (in this case mm2 db).

We would need a new global database for all mnemonics/seeds, since `MM2.db` path depends on the public key hash, which in turn depends on the seed. Please note that mm2 can be initialized with only one wallet/mnemonic at a time for now and that `MM2.db` or `MM2-shared.db` is specific to each wallet/mnemonic, and not shared across all wallets. The seed file is retrieved when starting mm2, based on the wallet name, there are no additional transactional operations for this file, except when using the `get_mnemonic` RPC.

We now have 2 options:

1 - Create a file for each wallet's encrypted mnemonic in the the db root directory. This can even allow us to support importing/exporting seeds to/from different wallets by supporting different file formats in the future, e.g. https://cryptobook.nakov.com/symmetric-key-ciphers/ethereum-wallet-encryption. This is why I mentioned the industry standards.
https://github.com/KomodoPlatform/komodo-defi-framework/blob/66085b00f6e3ecd68dfe40f4ed7694918fe8a30f/mm2src/mm2_core/src/mm_ctx.rs#L289-L293
2 - Create a global database in the db root directory that is only used when starting mm2 and for one RPC. If a user wants to import their seed file to another wallet without using mm2 they would have to access the database to get it.

I think option 1 is better :)